### PR TITLE
GH-125069: Fix inconsistent joining in `WindowsPath(PosixPath(...))`

### PR DESCRIPTION
**Pull request by barneygale**
`PurePath.__init__()` incorrectly uses the `_raw_paths` of a given `PurePath` object with a different flavour, even though the procedure to join path segments can differ between flavours.

This change makes the `_raw_paths`-enabled deferred joining apply _only_ when the path flavours match.

* Issue: gh-125069

### CODE REVIEW

File: Lib/test/test_pathlib/test_pathlib.py [hunk 1]
```diff
@@ -131,6 +131,15 @@ def test_constructor_nested(self):
         self.assertEqual(P(P('a'), P('b'), P('c')), P(FakePath("a/b/c")))
         self.assertEqual(P(P('./a:b')), P('./a:b'))
 
+    @needs_windows
+    def test_constructor_nested_foreign_flavour(self):
+        # See GH-125069.
```
Comment (rruuaanng) [inline Lib/test/test_pathlib/test_pathlib.py:136]: I prefer lowercase.
```
# See gh-125069.
```

File: Misc/NEWS.d/next/Library/2024-10-08-21-17-16.gh-issue-125069.0RP0Mx.rst [hunk 2]
```diff
@@ -0,0 +1,4 @@
+Fix issue where providing a :class:`pathlib.PurePath` object as an
+initializer argument to a second :class:`~pathlib.PurePath` object with a
+different :attr:`~pathlib.PurePath.parser` resulted in arguments to the
+former object's initializer being joined by the latter object's parser.
```
Comment (picnixz) [inline Misc/NEWS.d/next/Library/2024-10-08-21-17-16.gh-issue-125069.0RP0Mx.rst]:
```suggestion
Fix an issue where providing a :class:`pathlib.PurePath` object as an
initializer argument to a second :class:`~pathlib.PurePath` object with a
different :attr:`~pathlib.PurePath.parser` resulted in arguments to the
former object's initializer being joined by the latter object's parser.
```

**Comment by barneygale** [comment]
Some background info: the `_raw_paths`-enabled deferred joining is meant to speed up this sort of thing:

```python
str(Path('a') / 'b' / 'c' / 'd' / 'e')
```

Without deferred joining, this requires four calls to `os.path.join()`:

- `join('a', 'b')`
- `join('a/b', 'c')`
- `join('a/b/c', 'd')`
- `join('a/b/c/d', 'e')`

With deferred joining, it calls `join()` once: `join('a', 'b', 'c', 'd', 'e')`. This also saves three calls to `os.path.splitroot()`.

File: Lib/pathlib/_local.py [hunk 3]
```diff
@@ -119,9 +119,9 @@ def __init__(self, *args):
         paths = []
         for arg in args:
             if isinstance(arg, PurePath):
-                if arg.parser is ntpath and self.parser is posixpath:
+                if arg.parser is not self.parser:
                     # GH-103631: Convert separators for backwards compatibility.
```
Comment (picnixz) [inline Lib/pathlib/_local.py:123]: Is this comment still relevant?

Comment (barneygale) [reply Lib/pathlib/_local.py:123]: Yes - `as_posix()` converts separators.

**Review by picnixz — APPROVED** [review]
LGTM.

**Comment by barneygale** [comment]
Thank you very much for the review, @picnixz

**Comment by picnixz** [comment]
Always happy to review them!

**Pull request merged**